### ISSUE TITLE
Explore python libraries for direct docx export

### ISSUE THREAD

**Issue by holtchesley** [harvard-lil/h2o#1423]
We've run into a couple of issues with exports that might be fixed up with a switch to a different export library. 

1. Sometimes Pandoc runs out of memory for exports. 
2. Pandoc doesn't handle footnotes /perfectly/
  - We want to be able export footnotes or section notes. 
  - We'd like to be able to handle footnotes from a case, and from a casebook author with two different sets of symbols.

Additionally, we're considering:
 - Supporting other export formats
 - Supporting Templates for export

And want to understand how they might work.

**Comment by rebeccacremona** [comment]
Two libraries mentioned earlier today:

https://python-docx.readthedocs.io/en/latest/index.html doesn't presently handle footnotes at all; a number of partial PRs have been submitted, and it sounds like the maintainer is open to the idea: we might be able to adopt one of them and help it get merged in.

https://pypi.org/project/htmldocx/ is pretty basic: we might be able to use snippets of its code, but would need to use python-docx directly for sectioning, applying named styles, handling block quotes, and, I suspect, most non-`<p>` tags: there's hard-coded styling for `ol`/`li` and `<a>`, for instance, that I'm sure we don't want.

We would still need to build and insert our custom table of contents directly into the DOCX XML, as well as the bookmarks it points to.

The applying of custom styles (like `CaseBookSection` and `Ellision` and `BlockQuotes`) correctly, while traversing the tree, takes some doing, because word styles do not cascade like CSS: you apply the desired style to each block. See https://github.com/jgm/pandoc/pull/5670 for a discussion about precedence and accidentally clobbering nested styles.

**Comment by jcushman** [comment]
Copied from slack -- here’s my ancient code to convert an internal court manuscript docx to a formatted book-ready docx with python-docx, including some footnote manipulation ( https://github.com/harvard-lil/CAPNow/blob/develop/scripts/convert.py ), and here’s a helper file to dump differences between docx files, which is useful for debugging differences between your exported document and your test-suite-containing template docx ( https://github.com/harvard-lil/CAPNow/blob/develop/scripts/compare.py )

**Comment by jcushman** [comment]
Documenting my thoughts more on this ...

**Performance profiling**

I did some profiling of Pandoc back in 2019 (whatever version we were running at that point!). Back then I found that Pandoc export on the Lessig conlaw book took about 25 seconds. Our lua TOC script took about 5 seconds, and the remaining 20 seconds was consistent with the 750ms that the author reported for a document 1/20th as long, so I concluded that we couldn't do much on the pandoc side to improve performance.

At the same time I did some quick profiling of a pure-python method, by parsing the full casebook html prepared for pandoc with lxml, stepping through each `<p>` tag and adding it to reference.docx as a text string with python-docx. (Which I felt was on the order of the amount of python runtime that would be necessary to export a correct book, to get a feel for performance before doing that work.) That benchmark took about 4 seconds, or 15% the time of the pandoc version. So I concluded that a pure-python version could be much faster. I'd hazard that this approach also uses less RAM, but I didn't profile that.

I also did some experiments in parallelizing pandoc export, and found that it does parallelize performantly (a book 1/10th as long takes 1/10th as long to convert), so with enough CPU cores one could pandoc each book section separately and paste them back together in python-docx with some fixed overhead. That struck me as less clean than a pure-python approach, because maintaining that pipeline requires keeping more technologies in your head, but it could work.

**Footnotes etc.**

A separate thread is whether a pure-python approach could help improve output quality. Philosophically I think we'll bring the most joy with our export button if we start by building an ideal model document in our docx template and then work backwards to design an export pipeline. One way or another I'd like to find a set of export steps that can handle Word-native footnote numbering and anything else that's allowed in docx and useful to our authors.

On this side I have the experience of writing CAPNow, linked in the previous comment, where my goal was to convert an unstyled docx case manuscript into a print-ready book with page headers and so on, that would be acceptable to the court as their print-on-demand book output. My memory is that python-docx made it _possible_ to do anything, which makes sense because docx is zipped xml files and python-docx is a set of helpers to make it easier to edit the xml files. Some parts were easier than others, though -- when python-docx didn't support part of the spec, like footnotes, I would load the raw xml and rewrite it with PyQuery. So the code ends up conceptually confusing as far as moving between objects that represent python-docx's higher level API, and objects that represent PyQuery editors for the docx's raw XML. Looking back at my research-quality code for CAPNow five years later, it would definitely need more commenting/scaffolding to make sense.

As far as development experience, the way I worked on CAPNow was to create a model docx with all the features I needed to support, then use a diff tool to compare my code's current output to the model docx, with any diff being a thing that needed fixing, then drop an ipdb.set_trace() at the point in my code that was supposed to generate the missing xml and iterate in pdb to find what generated the correct thing. That turned out to be a very usable development cycle to hone in on the xml I needed to generate.

A particular challenge with this approach is supporting the full range of HTML in H2O cases -- i.e. going from "stepping through each `<p>` tag and adding it to reference.docx as a text string" in my benchmark to actually emitting the correct docx xml for italics and tables and whatever's in there. I could see a few plausible approaches here, like using or adapting an existing library like htmldocx, or looping through each element in raw lxml, or using a SAX parser, or even calling out to pandoc for a first pass (below). Would be inclined to do some exploratory programming and see what gets closest to a full solution with least effort.

**More on the stitching method**

As far as the philosophical goal of making a pipeline that can make arbitrary docx, another viable design besides pure python-docx would be to scatter out each case or section to pandoc for processing, then have a post-processing step in python-docx. That could include TOC-building (which wins back some runtime and multi-language complexity from lua), and also footnote post-processing or any other docx-specific stuff that we want, if we inject something like bracketed formatting tags that will be ignored by pandoc and then picked up by our post-processing code. Or we could keep the full existing pandoc pipeline and just do a new cleanup pass in python-docx.

This might have the advantage that it would preserve some of pandoc's ability to make different formats? Like maybe it's easier to switch to making an epub in this model. Has the disadvantage that a large chunk of the pipeline stays as a black box that's slower and tricky to modify, though. I dunno, this is where I'd get to exploratory programming to see how it feels.
